### PR TITLE
HttpResponseMessage.ReasonPhrase should not be null

### LINKED ISSUE #6721: System.Net.Http.HttpResponseMessage.ReasonPhrase can return null on core
There was a recent regression with WinHttpHandler where HttpResponseMessage.ReasonPhrase now returns `null` rather than `string.Empty` if no reason phrase was specified in the response message. This diverges from the HttpClientHandler/HttpWebRequest behavior in the full framework that always return `string.Emtpy`. 

@davidsh

### PR DESCRIPTION
**Pull request by davidsh**
Recent changes due to PR #3199 caused the `HttpResponseMessage.ReasonPhrase` property to be a null string if the HTTP server response only contained a statuscode but no description on the status line. The existing .NET Framwork behavior is to have empty string as the property value in that case.

Added a new test to validate this change.

Fixes #6721.

### CODE REVIEW

**Comment by davidsh** [comment]
@stephentoub @justinvp PTAL

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs [hunk 1]
```diff
@@ -51,6 +51,10 @@ internal static class WinHttpResponseParser
             {
                 response.ReasonPhrase = GetReasonPhrase(response.StatusCode, buffer, reasonPhraseLength);
             }
+            else
+            {
+                response.ReasonPhrase = string.Empty;
+            }
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: Nit: if it were me, I'd have done:
```C#
response.ReasonPhrase = responsePhraseLength > 0 ?
    GetReasonPhrase(response.StatusCode, buffer, reasonPhraseLength) :
    string.Empty;
```
but what you have is fine.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 2]
```diff
@@ -696,20 +696,20 @@ public static IEnumerable<object[]> VerifyUploadServersStreamsAndExpectedData
             }
         }
 
-        [Fact]
-        public async Task GetAsync_CallMethod_ExpectedStatusLine()
+        [Theory]
+        [InlineData(HttpStatusCode.MethodNotAllowed, "Custom description")]
+        [InlineData(HttpStatusCode.MethodNotAllowed, "")]
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Sometimes [Theory]s just make me smile :smile:

**Comment by stephentoub** [comment]
LGTM

**Comment by justinvp** [comment]
Thanks for the fix, @davidsh. I was going to offer to submit a fix with a temporary test case using the loopback server, but you are too fast!

The changes LGTM getting back to the previous behavior of `WinHttpHandler`, but is it the correct behavior?

Recently, I have been comparing the behavior of `WinHttpHandler` and `CurlHandler` against each other and against `HttpClient` (and `HttpWebRequest`) on the full framework and have found a number of discrepancies around the status line and header parsing. (I plan to submit a PR with tests and fixes to consider, unfortunately I haven't had time to pull it together yet).

Consider the following tests run on the full framework (.NET 4.6). All the cases pass except the one noted in the comment.

With `WinHttpHandler` in .NET Core, I'd expect `"HTTP/1.1 200"` to return `"OK"` to match the full framework, not `string.Empty`.

```c#
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
using Xunit;

public class Tests
{
    public static IEnumerable<object[]> StatusLineData
    {
        get
        {
            yield return new object[] { "HTTP/1.1 200 OK", "OK" };
            yield return new object[] { "HTTP/1.1 200", "OK" }; // HttpWebRequest test fails: it returns null instead of expected "OK"
            yield return new object[] { "HTTP/1.1 200 ", "" };
            yield return new object[] { "HTTP/1.1 200  ", " " };
            yield return new object[] { "HTTP/1.1 200   ", "  " };
            yield return new object[] { "HTTP/1.1 299 FOO", "FOO" };
            yield return new object[] { "HTTP/1.1 299", null };
        }
    }

    [Theory]
    [MemberData(nameof(StatusLineData))]
    public async Task HttpClient_ValidStatusLine_SuccessfullyParsed(string statusLine, string expectedReasonPhrase)
    {
        IPEndPoint serverEndPoint;
        Task serverTask = LoopbackServer.StartServer(out serverEndPoint, statusLine);

        using (var client = new HttpClient())
        {
            var uri = new Uri($"http://{serverEndPoint.Address}:{(serverEndPoint).Port}/");
            using (var response = await client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead))
            {
                Assert.Equal(expectedReasonPhrase, response.ReasonPhrase);
            }
        }

        await serverTask;
    }

    [Theory]
    [MemberData(nameof(StatusLineData))]
    public async Task HttpWebRequest_ValidStatusLine_SuccessfullyParsed(string statusLine, string expectedReasonPhrase)
    {
        IPEndPoint serverEndPoint;
        Task serverTask = LoopbackServer.StartServer(out serverEndPoint, statusLine);

        var uri = new Uri($"http://{serverEndPoint.Address}:{(serverEndPoint).Port}/");
        var request = (HttpWebRequest)WebRequest.Create(uri);

        using (var response = await request.GetResponseAsync())
        {
            var httpResponse = (HttpWebResponse)response;
            Assert.Equal(expectedReasonPhrase, httpResponse.StatusDescription);
        }

        await serverTask;
    }
}

public static class LoopbackServer
{
    private static readonly Encoding s_defaultHttpEncoding = Encoding.GetEncoding("iso-8859-1");

    public static Task StartServer(out IPEndPoint serverEndPoint, string statusLine)
    {
        var server = new TcpListener(IPAddress.Loopback, 0);
        Task serverTask = ((Func<Task>)async delegate {
            server.Start();
            using (var client = await server.AcceptSocketAsync())
            using (var stream = new NetworkStream(client))
            using (var reader = new StreamReader(stream, Encoding.ASCII))
            {
                // Read past request headers.
                string line;
                while (!string.IsNullOrEmpty(line = reader.ReadLine())) ;

                string content = "This is some response content.";

                // Write response.
                using (var writer = new StreamWriter(stream, s_defaultHttpEncoding))
                {
                    writer.Write($"{statusLine}\r\n");
                    writer.Write($"Date: {DateTimeOffset.UtcNow:R}\r\n");
                    writer.Write("Content-Type: text/plain\r\n");
                    writer.Write("\r\n");
                    writer.Write($"{content}\r\n");
                    writer.Flush();
                }

                client.Shutdown(SocketShutdown.Both);
            }
        })();

        serverEndPoint = (IPEndPoint)server.LocalEndpoint;
        return serverTask;
    }
}
```

**Comment by davidsh** [comment]
>The changes LGTM getting back to the previous behavior of WinHttpHandler, but is it the correct behavior?
Recently, I have been comparing the behavior of WinHttpHandler and CurlHandler against each other and against HttpClient (and HttpWebRequest) on the full framework and have found a number of discrepancies around the status line and header parsing. (I plan to submit a PR with tests and fixes to consider, unfortunately I haven't had time to pull it together yet).

This is very interesting. In addition to testing again the .NET Framework (Desktop), we also need to test again the UWP version of System.Net.Http which uses WinRT Windows.Web.Http underneath. I don't think we have a full test matrix of this and it will interesting to see what differences we have.

At a minimum, I would like to at least get this PR in to get back to the behavior we had before your change. But I would like this topic further pursued with tests across all the platforms to see where w we land.

Usually in terms of "System.Net.Http" contract behavior,we prefer to align with the .NET Framework (Desktop) behavior as much as possible, even if there are cases which might appear as "inconsistent" w.r.t. different status codes having different default values for missing reasonphrases on the status line. But I'd like to see data on this before making any decisions.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs (hunk 1 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs]: Fixed.

**Comment by stephentoub** [comment]
LGTM

**Pull request merged**